### PR TITLE
VC operations

### PR DESCRIPTION
**Pull request by luigiw**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/ml/azure-ai-ml/azure/ai/ml/_utils/azure_reource_utils.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+# ---------------------------------------------------------
```
Comment (diondrapeck) [inline sdk/ml/azure-ai-ml/azure/ai/ml/_utils/azure_reource_utils.py:1]: typo in file name - *azure_resource_utils.py

File: sdk/ml/azure-ai-ml/azure/ai/ml/_utils/azure_reource_utils.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    # Create query
+    argQuery = arg.models.QueryRequest(subscriptions=subsList, query=strQuery, options=argQueryOptions)
+
+    # Allowing API version to be set is yet to be released by azure-mgmt-resourcegraph
+    # return argClient.resources(argQuery, api_version="2021-03-01") # This is the API version Studio UX is using, but why does it matter
+    return argClient.resources(argQuery)
+
+
+def get_vitual_clusters_from_subscriptions(credential: TokenCredential, subscription_list: Optional[List[str]] = None) -> List[Dict]:
```
Comment (diondrapeck) [inline sdk/ml/azure-ai-ml/azure/ai/ml/_utils/azure_reource_utils.py:34]: typo: *get_virtual_clusters_from_subscriptions

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_virtual_cluster_operations.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+        if scope is None:
+            subscription_list = None
+        elif scope.lower() == Scope.SUBSCRIPTION:
+            subscription_list = [self._subscription_id]
+        else:
+            message = "Invalid scope: {}. Valid values are 'subscription' or None.".format(scope)
+            raise UserErrorException(message=message, no_personal_data_message=message)
+
+        return get_vitual_clusters_from_subscriptions(self._credentials, subscription_list=subscription_list)
```
Comment (diondrapeck) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_virtual_cluster_operations.py:59]: same typo from before

**Review by diondrapeck — CHANGES_REQUESTED** [review]

File: sdk/ml/azure-ai-ml/azure/ai/ml/_utils/azure_reource_utils.py (hunk 2 above)
Comment (luigiw) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_utils/azure_reource_utils.py]: Good catch. Fixed.

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_virtual_cluster_operations.py (hunk 3 above)
Comment (luigiw) [reply sdk/ml/azure-ai-ml/azure/ai/ml/operations/_virtual_cluster_operations.py]: Fixed.

**Comment by luigiw** [comment]
@kdestin test

File: sdk/ml/azure-ai-ml/setup.py [hunk 4]
```diff
@@ -85,6 +85,8 @@
         "azure-common<2.0.0,>=1.1",
         "typing-extensions<5.0.0",
         # "opencensus-ext-azure<2.0.0", disabled until SDK logging re-activated
+        "azure-mgmt-resourcegraph>=2.0.0",
```
Comment (kdestin) [inline sdk/ml/azure-ai-ml/setup.py:88]: Do these requirements need an upper bound?

File: sdk/ml/azure-ai-ml/tests/virtual_cluster/e2etests/test_vc.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+@pytest.mark.timeout(600)
+@pytest.mark.usefixtures(
+    "recorded_test"
+)
+@pytest.mark.training_experiences_test
+class TestVirtualCluster(AzureRecordedTestCase):
+
+    @pytest.mark.e2etest
+    def test_get_and_list(self, client: MLClient) -> None:
```
[the remark this answers is not in the document]
Comment (luigiw) [reply sdk/ml/azure-ai-ml/tests/virtual_cluster/e2etests/test_vc.py:13]: Some code was missed out in the previous commit. I committed the "get" part now.

File: sdk/ml/azure-ai-ml/tests/conftest.py [hunk 6]
```diff
@@ -832,6 +832,7 @@ def pytest_configure(config):
         ("data_experiences_test", "marks tests as data experience tests"),
         ("local_endpoint_local_assets", "marks tests as local_endpoint_local_assets"),
         ("local_endpoint_byoc", "marks tests as local_endpoint_byoc"),
+        ("vc_test", "marks tests as virtual cluster tests"),
```
[the remark this answers is not in the document]
Comment (luigiw) [reply sdk/ml/azure-ai-ml/tests/conftest.py]: Good suggestion. Changed to `virtual_cluster_test`.

File: sdk/ml/azure-ai-ml/azure/ai/ml/_utils/azure_resource_utils.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    return get_resources_from_subscriptions(strQuery, credential, subscription_list).data
+
+
+def get_generic_resource_by_id(arm_id: str, credential:  TokenCredential, subscription_id: str, api_version: Optional[str] = None) -> Dict:
+    resource_client = ResourceManagementClient(credential, subscription_id)
+    generic_resource = resource_client.resources.get_by_id(arm_id, api_version)
+
+    return generic_resource.as_dict()
+
+def get_vc_by_id(name: str, resource_group: str, subscription_id: str, credential: TokenCredential) -> Dict:
```
[the remark this answers is not in the document]
Comment (luigiw) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_utils/azure_resource_utils.py]: Done.

File: sdk/ml/azure-ai-ml/setup.py (hunk 4 above)
Comment (kdestin) [reply sdk/ml/azure-ai-ml/setup.py]: Resolved in #28624

**Pull request closed without merging**